Decorated instructor dashboard with sudo_required.

waheedahmed opened the pull request:
[PLAT-407] (https://openedx.atlassian.net/browse/PLAT-407), [PLAT-394] (https://openedx.atlassian.net/browse/PLAT-394)
Decorated instructor dashboard with sudo_required and added couple of unit tests.
Decorated studio permissions with sudo_required.

waheedahmed commented:
@cpennington, @adampalay please review this.

waheedahmed commented:
test this please

waheedahmed commented:
test please

waheedahmed commented:
test this please

waheedahmed commented:
test this please

waheedahmed commented:
@cpennington please take a look at this, I fixed all unit tests, bokchoy and feature tests which were failing due to ```sudo_required``` decorator.

adampalay commented:
@waheedahmed now you have to rebase :(

cpennington commented on cms/djangoapps/contentstore/views/tests/test_user.py:
```diff
@@ -14,6 +14,7 @@
 class UsersTestCase(CourseTestCase):
     def setUp(self):
         super(UsersTestCase, self).setUp()
+        self.grant_sudo_access(self.course.id.to_deprecated_string(), self.user_password)
```
`unicode(self.course.id)`, rather than `.to_deprecated_string`

cpennington commented on cms/djangoapps/contentstore/views/user.py:
```diff
@@ -38,6 +39,7 @@ def request_course_creator(request):
 @login_required
 @ensure_csrf_cookie
 @require_http_methods(("GET", "POST", "PUT", "DELETE"))
+@sudo_required
```
Does the Backbone app that calls this view correctly prompt to get sudo access when the user hits this view?

waheedahmed replied:
@cpennington nope but i will add this too.

cpennington commented on common/djangoapps/django_sudo/decorators.py:
```diff
@@ -0,0 +1,56 @@
+    >>> @sudo_required('admin_page')
+    >>> def secure_admin_page(request):
+    >>>     ...
+    """
```
This docstring should mention the addition of `course_id`/`course_key_string`/`library_key_string` parsing to set the region.

cpennington commented on common/djangoapps/edx_admin/tests.py:
```diff
@@ -0,0 +1,16 @@
+        """
+        Tests that 1 + 1 always equals 2.
+        """
+        self.assertEqual(1 + 1, 2)
```
You can delete this file.

cpennington commented on lms/djangoapps/courseware/tests/test_view_authentication.py:
```diff
@@ -273,6 +287,8 @@ def test_global_staff_access(self):
+        self.grant_sudo_access(self.test_course.id.to_deprecated_string(), 'test')
         # and now should be able to load both
         urls = [reverse('instructor_dashboard', kwargs={'course_id': self.course.id.to_deprecated_string()}),
                 reverse('instructor_dashboard', kwargs={'course_id': self.test_course.id.to_deprecated_string()})]
```
It would be good to add a test that without sudo access, you can't access the instructor dashboard.

cpennington commented on lms/djangoapps/instructor/tests/test_api.py:
```diff
@@ -3215,6 +3229,7 @@ def setUp(self):
+        self.grant_sudo_access(self.course.id.to_deprecated_string(), 'test')
 
     def test_change_due_date(self):
         url = reverse('change_due_date', kwargs={'course_id': self.course.id.to_deprecated_string()})
```
It would be good to add some tests that verify that the api endpoints aren't available without sudo mode.

cpennington commented:
Assuming reasonable answers to my comments, :+1:

waheedahmed commented:
@cpennington I will address all your comment's but I have question about ```sudo.html``` files for LMS and CMS, in LMS you used ```main_django.html``` to add header and footers but on studio side mako template language is used and when i add ```base.html``` in ```sudo.html``` it is not rendered correctly, so how i add this because sudo template is rendered by django not mako?

cpennington commented:
`main_django.html` exists in the LMS to support exactly this use case. We may just need to add a similar django-ified copy of `base.html` on the studio side.

waheedahmed commented:
test this please

waheedahmed commented:
test this please

waheedahmed commented:
test this please

waheedahmed commented:
test this please

waheedahmed commented:
@cpennington @adampalay I created a sandbox for ```django-sudo``` http://django-sudo.m.sandbox.edx.org.

adampalay commented:
@explorerleslie ^^

explorerleslie commented:
@adampalay what am I supposed to respond to on this PR? I have no idea what it means.

waheedahmed commented:
test this please

waheedahmed commented:
test this please

cpennington commented on common/djangoapps/xmodule_modifiers.py:
```diff
@@ -277,6 +277,7 @@ def add_staff_markup(user, has_instructor_access, block, view, frag, context):
                      'category': str(block.__class__.__name__),
                      # Template uses element_id in js function names, so can't allow dashes
                      'element_id': block.location.html_id().replace('-', '_'),
+                     'region': unicode(course_id).replace(':', '').replace('/', '_').replace('+', '_'),
```
It's a bit unfortunate that this replacement is done twice (once here, and one in the `sudo_required` decorator). Could this simply pull the active region? Or could you extract the function that maps from `CourseKey` to region?

waheedahmed replied:
@cpennington there is no active region here we need this for sudo modal to submit user password to get sudo access.
I will create a function to map ```CourseKey``` to region.

cpennington commented on lms/djangoapps/courseware/tests/test_view_authentication.py:
```diff
@@ -387,6 +403,26 @@ def test_enrollment_period(self):
+        self.login(self.org_staff_user)
+        url = reverse('instructor_dashboard', kwargs={'course_id': unicode(self.course.id)})
+        response = self.assert_request_status_code(302, url)
+        self.assertIn('sudo/?region=', response['location'])
```
Optional: It might be worth testing that the region redirected to is the correct one, based on the course id

cpennington replied:
(You might be able to do that by having django revers construct the correct url for the redirection.

cpennington commented on lms/static/js/staff_debug_actions.js:
```diff
@@ -127,4 +136,8 @@ $(document).ready(function() {
+    $courseContent.on("click", '.staff-debug', function() {
+        StaffDebug.check_sudo_and_view_modal($(this).data('element-id'), $(this).data('is-sudo'));
+        return false;
+    });
```
Is this trying to enforce sudo mode for the staff-debug information? If so:

1. it doesn't look secure, since it's relying on a client-side check that a) sudo mode is active, and b) what should happen based on the sudo mode. Both of these can be spoofed by the end-user
2. sudo mode should never be active when looking at courseware

waheedahmed replied:
@cpennington This is for staff-debug-info, but it doesn't mean that if user change its value manually and then he can do staff user stuff it will give them a error on doing that, it is only for showing the modal.

cpennington replied:
I don't think that the staff-debug view should be protected by sudo-mode, because the staff-debug information is displayed inside the courseware (and even staff users shouldn't have sudo-mode activated when viewing courseware).

waheedahmed commented:
test this please

waheedahmed commented on lms/templates/staff_problem_info.html:
```diff
@@ -64,12 +64,10 @@
-        <a href="#" class="staff-debug-sdelete">${_('Delete Student State')}</a>
-        |
-        <a href="#" class="staff-debug-rescore">${_('Rescore Student Submission')}</a>
+            <a href="#" class="staff-debug-grade-adjustments">${_('Reset Student Attempts | Delete Student State | Rescore Student Submission')}</a>
```
@mhoeber Please review this and suggest a text string for this link which can work for both cases.

mhoeber replied:
Suggest:  "Modify Student's State for Problem"

mhoeber commented on cms/templates/sudo/sudo.html:
```diff
@@ -0,0 +1,48 @@
+<section aria-hidden="true" class="modal sudo-modal" id="sudo-modal" style="overflow:auto; display: none;" >
+  <div class="inner-wrapper" style="color:black">
+    <header>
+      <h2>{% trans "Confirm Your Password To Get Sudo Access" %}</h2>
```
Change heading to:  Confirm Your Password to Access Instructor Dashboard

mhoeber commented on lms/templates/sudo/sudo.html:
```diff
@@ -0,0 +1,21 @@
+<section style="margin: 0 auto; width: 500px; padding: 50px;">
+    <div class="inner-wrapper">
+        <header>
+          <h2>{% trans "Confirm Your Password To Get Sudo Access" %}</h2>
```
Change heading to: Confirm Your Password to Access the Instructor Dashboard

cpennington commented on common/djangoapps/django_sudo/utils.py:
```diff
@@ -0,0 +1,32 @@
+import sudo.middleware
+
+
+def region_name(region):
```
Should this just be an internal function of the `django_sudo` 3rd party library to sanitize region names? The fact that it's being stored as a cookie is part of the `django_sudo` code, right?

waheedahmed replied:
@cpennington No, it is for edx specific ```course_id``` region to make it valid for cookie. we are just removing these three character's ```[':', '/', '+']``` from ```course_id``` in this function.

cpennington replied:
Our region is edx-specific, but the operation to make the region safe to put in a cookie isn't edx-specific. If you moved this functionality to nearer to the cookie storage, then a) it would be clearer, and b) you wouldn't need to call it every time you specify a region.

cpennington replied:
I'd still like an answer to this. It doesn't seem like the act of making the region name safe to store in a cookie is edx specific code.

cpennington commented:
:+1:

waheedahmed commented:
test this please

waheedahmed commented:
test this please

waheedahmed commented:
test this please

waheedahmed commented:
@cpennington Final review please.

waheedahmed commented:
@explorerleslie Can you please tag someone from TNL who you think should review this?

cpennington commented on common/djangoapps/edx_admin/admin.py:
```diff
@@ -0,0 +1,29 @@
+    """
+    django_autodiscover()
+    for model, modeladmin in django_site._registry.items():     # pylint: disable=protected-access
+        if model not in site._registry:                         # pylint: disable=protected-access
```
Nit: You could put this pylint disable at the top of the block so that you only need it once.

cpennington commented:
:+1:

explorerleslie commented:
@cahrens a review for this PR is requested by @waheedahmed.

cahrens commented:
@waheedahmed Can you explain what this PR is doing and how sudo_required works? The Platform tickets mentioned in the summary have no details.

waheedahmed commented:
@cahrens This PR is implementing a sudo-mode to prevent against malicious course authors hijacking administrative rights. This helps prevent us against malicious javascript a course author might put in an Xblock.

It's a feature that has a user with elevation permissions enter in their password periodically when doing things that require elevated permissions--accessing the instructor dashboard or the django admin panel and course team page on studio side.

If instructor/staff user visit the instructor dashboard or course team page (studio side) ```sudo_required``` will redirect them to a sudo page to confirm their password to access these views.

```sudo_required``` needs a region name which can be None or a keyword and we are using ```course_id``` as a region here to give user sudo access only for that course which actually he/she is accessing.

And when user leaves that page we are revoking it's sudo access.

cahrens commented:
@waheedahmed please link a working sandbox

waheedahmed commented:
@cahrens Sure I'll add it to the TNL review ticket.

waheedahmed commented:
test this please

waheedahmed commented:
@cpennington addressed you comments, does it looks good now?

waheedahmed commented:
test this please

waheedahmed commented:
test this please

dan-f commented:
Hi @waheedahmed. I'm going to do a review from Teaching and Learning.

dan-f commented on lms/djangoapps/instructor/tests/test_api.py:
```diff
@@ -308,6 +309,26 @@ def test_staff_level(self):
+        self.client.login(username=staff_member.username, password='test')
+
+        for endpoint, args in self.staff_level_endpoints:
+            # # TODO: make these work
```
Is this TODO still relevant? Why don't they work?

dan-f commented on lms/djangoapps/instructor/tests/test_api.py:
```diff
@@ -340,6 +362,37 @@ def test_instructor_level(self):
+        self.client.login(username=inst.username, password='test')
+
+        for endpoint, args in self.staff_level_endpoints:
+            # TODO: make these work
```
Same question here.

dan-f commented on lms/djangoapps/instructor/tests/test_api.py:
```diff
@@ -340,6 +362,37 @@ def test_instructor_level(self):
+            )
+
+        for endpoint, args in self.instructor_level_endpoints:
+            # TODO: make this work
```
And here.

dan-f commented on common/djangoapps/django_sudo/decorators.py:
```diff
@@ -0,0 +1,55 @@
+"""
```
High level question - why is this package named `django_sudo`? I'm finding this confusing, since the library itself is called `django-sudo`. Would it be alright to call this something along the lines of `django_sudo_helpers`?

waheedahmed replied:
@dan-f I named it ```django_sudo``` because I wrote a custom ```sudo_required``` decorator instead of using default.
Yeah we can change it to ```django_sudo_helpers```.

dan-f commented on lms/djangoapps/instructor/views/instructor_dashboard.py:
```diff
@@ -65,8 +66,33 @@ def is_enabled(cls, course, user=None):  # pylint: disable=unused-argument,redef
         return user and has_access(user, 'staff', course, course.id)
 
 
+def check_staff_or_404():
```
Is this a new security check, independent of `sudo_required`? I'm surprised that this is not already implemented across both instructor dashboards.

dan-f replied:
On that note - what about the old instructor dashboard?

waheedahmed replied:
@dan-f There is no check like that in old instructor dashboard. I just refactored this due to a unit test failure.

dan-f commented on lms/djangoapps/instructor/views/instructor_dashboard.py:
```diff
@@ -399,11 +425,21 @@ def _is_small_course(course_key):
 
 
-def _section_student_admin(course, access):
+def _section_student_admin(course, access, unique_student_identifier, problem_to_reset):
```
I'm having trouble understanding the changes in this function. What does this have to do with `sudo_required`?

waheedahmed replied:
@dan-f I removed all three action links on ```staff_debug_info``` modal after talking to @explorerleslie and @cpennington and made a single link which will redirect the user to instructor dashboard and auto populate the username and problem id in section and we need to get the problem link back to courseware that's why I made this change to return all these values in ```section_data```.

dan-f replied:
Ah got it! Thanks for the explanation.

dan-f commented:
@waheedahmed - I'm seeing some unexpected behavior on the instructor dashboard. When I click on the "List enrolled students' profile information" button, I see a quick DOM change that gets reverted so fast that I can't tell what happened. The network tab shows that I got redirected to the sudo view, but the page does not actually load. Upon further testing, the same thing happens for every button on this page.

Furthermore, the "list_report_downloads" endpoint is similarly returning redirects that the browser isn't responding to.

dan-f commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -261,6 +262,9 @@ def course_handler(request, course_key_string=None):
             if course_key_string is None:
                 return redirect(reverse("home"))
             else:
+                # Revoke sudo privileges from a request explicitly
```
Is the goal to revoke sudo privileges any time a user navigates away from a `sudo_required` page? If so, this particular implementation seems like an incomplete solution to me, as we will never know where a course author/administrator will go after viewing such a page. Isn't that what the `SUDO_COOKIE_AGE` setting is for?

cpennington replied:
We should be revoking sudo privileges before we display any XBlock.

dan-f replied:
So why are we doing this on the `course_handler`, which doesn't render XBlocks?

cpennington replied:
That's a good question. @waheedahmed, have any input on that?

waheedahmed replied:
@dan-f Sorry for that I thought may be we need to revoke sudo privileges on course load too, I'll remove that.

dan-f commented on common/djangoapps/django_sudo/utils.py:
```diff
@@ -0,0 +1,15 @@
+    Initialize the session and is_sudo on request object.
+    """
+    session_middleware = django.contrib.sessions.middleware.SessionMiddleware()
+    session_middleware.process_request(request)
```
Is it necessary to explicitly call this? `django.contrib.sessions.middleware.SessionMiddleware` is included in `MIDDLEWARE_CLASSES` in both Studio and LMS.

cpennington replied:
It looks like it's used in tests, but given that, I'd recommend moving the file to `django_sudo/test/utils.py`

waheedahmed replied:
@cpennington, @dan-f  I'll do that.

dan-f commented on common/test/acceptance/pages/lms/instructor_dashboard.py:
```diff
@@ -773,23 +774,23 @@ def is_browser_on_page(self):
 
     @property
-    def student_email_input(self):
+    def ee_student_email_input(self):
```
Nit: `ee` is not a helpful prefix. Why not use `entrance_exam` instead?

waheedahmed replied:
@dan-f When i changed ```ee``` to ```entrance_exam``` pylint giving me ```invalid-name``` error.

dan-f commented on common/test/acceptance/tests/studio/base_studio_test.py:
```diff
@@ -152,3 +153,11 @@ def log_in(self, user, is_staff=False):
         )
         auth_page.visit()
+
+    def get_sudo_access(self, redirect_page, password):
```
Please import this from helpers rather than re-define it here.

dan-f commented on cms/djangoapps/contentstore/views/library.py:
```diff
@@ -60,6 +62,9 @@ def library_handler(request, library_key_string=None):
 
     # request method is get, since only GET and POST are allowed by @require_http_methods(('GET', 'POST'))
     if library_key_string:
+        # Revoke sudo privileges from a request explicitly
```
Same question - why revoke here?

waheedahmed replied:
@dan-f Good catch, I did it at wrong place we need to revoke it in ```_display_library``` instead here because in ```_display_library```  we are displaying xblocks.

dan-f commented on lms/djangoapps/instructor/tests/test_legacy_xss.py:
```diff
@@ -39,6 +39,9 @@ def setUp(self):
             course_id=self._course.id
         )
 
+    def _is_sudo(self, region=None):    # pylint: disable=unused-argument
```
Why is this single-use function declared as a method? Could you re-define it as an inline lambda? It's confusing for this named method to always return `True`.

dan-f commented on lms/djangoapps/instructor/views/api.py:
```diff
@@ -1047,6 +1057,7 @@ def re_validate_invoice(obj_invoice):
 @cache_control(no_cache=True, no_store=True, must_revalidate=True)
 @require_level('staff')
+@sudo_required
 def get_students_features(request, course_id, csv=False):  # pylint: disable=redefined-outer-name
```
As I wrote earlier, I'm afraid that decorating all these functions with `sudo_required` won't necessarily yield the results you expect.

waheedahmed replied:
@dan-f Yeah I know that but we need to sure about that when someone do an ajax request without sudo privileges he should redirected to sudo page.

As we are giving user a sudo privileges when he/she visits instructor dashboard so it will not hurt us.

dan-f replied:
I should have clarified - I think that requiring sudo for all these commands probably makes sense. But sudo privileges are revoked on a timer, and once they have been revoked, the UX is confusing - my browser doesn't follow a redirect from these ajax requests.

dan-f commented on lms/static/coffee/src/instructor_dashboard/student_admin.coffee:
```diff
@@ -64,6 +64,11 @@ class @StudentAdmin
     @$request_response_error_ee       = @$section.find ".entrance-exam-grade-container .request-response-error"
     @$request_response_error_all    = @$section.find ".course-specific-container .request-response-error"
 
+    $student_grade_container = find_and_assert @$section, ".student-grade-container"
```
What does this have to do with the `django_sudo` changes?

waheedahmed replied:
@dan-f As i mentioned above we are populating username and problem_id in instructor dashboard and scrolling user down to that section when user click on link in ```staff_debug_modal```.

dan-f replied:
Yes, thanks again for the explanation.

dan-f commented on lms/static/js/staff_debug_actions.js:
```diff
@@ -17,95 +19,21 @@ var StaffDebug = (function(){
     return uname;
-  }
-
-  do_idash_action = function(action){
```
I'm having a hard time picking apart what's going on in this file, and how your changes relate to that. Why have you removed all these methods?

waheedahmed replied:
@dan-f We removed all three ```staff_debug_info``` links and made a single link that's why I removed these functions and added a single link functionality here.

dan-f commented on lms/templates/staff_problem_info.html:
```diff
@@ -67,15 +70,7 @@
-        <a href="#" class="staff-debug-rescore">${_('Rescore Student Submission')}</a>
-        % endif
-        ]
+        [<a href="#" class="staff-debug-grade-adjustments">${_("Modify Student's State for Problem")}</a>]
```
Has this change been reviewed by documentation/product? @explorerleslie

waheedahmed replied:
@dan-f we changed this after getting review from @mhoeber.

dan-f commented on lms/templates/sudo/sudo.html:
```diff
@@ -0,0 +1,21 @@
+<section style="margin: 0 auto; width: 480px; padding: 50px;">
+    <div class="inner-wrapper">
+        <header>
+          <h2 style="text-align: center;">{% trans "Confirm Your Password to Access the Instructor Dashboard" %}</h2>
```
I didn't realize that we support standard django templates (usually we use mako). @sarina - can you confirm whether `trans` actually works? I've never seen this before in the platform.

sarina replied:
Dan, did you simply try searching for this in the codebase? And also just
testing to see if it works?
http://edx.readthedocs.org/projects/edx-developer-guide/en/latest/internationalization/i18n.html#coverage-testing
- I'm a big fan of testing this out to actually see if it works.

On Tue, Jul 7, 2015 at 2:11 PM, Daniel Friedman <notifications@github.com>
wrote:

> In lms/templates/sudo/sudo.html
> <https://github.com/edx/edx-platform/pull/7404#discussion_r34071271>:
>
> > @@ -0,0 +1,21 @@
> > +{% extends "main_django.html" %}
> > +{% load i18n %}
> > +
> > +{% block body %}
> > +<section style="margin: 0 auto; width: 480px; padding: 50px;">
> > +    <div class="inner-wrapper">
> > +        <header>
> > +          <h2 style="text-align: center;">{% trans "Confirm Your Password to Access the Instructor Dashboard" %}</h2>
>
> I didn't realize that we support standard django templates (usually we use
> mako). @sarina <https://github.com/sarina> - can you confirm whether trans
> actually works? I've never seen this before in the platform.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/7404/files#r34071271>.
>

dan-f replied:
Good call. This is what I'm seeing

![screen shot 2015-07-07 at 2 28 56 pm](https://cloud.githubusercontent.com/assets/2146312/8553928/bd740306-24b4-11e5-8bc8-9a234e6f8c60.png)


Does @waheedahmed need to manually run `paver i18n_dummy` on the sandbox, or is that done already during provisioning?

dan-f commented:
@waheedahmed I'm seeing another unexpected behavior. I can manually go to `edx-domain/sudo`(for example, http://django-sudo.m.sandbox.edx.org/sudo/) without being redirected to that page. @explorerleslie is that what we want?

dan-f commented:
@waheedahmed I'm done with my first round of review. Thanks!

waheedahmed commented:
@dan-f I can't reproduce the above issue on instructor dashboard data download section, network tab is not showing any redirection to sudo.
May be your django-sudo session was expired.

waheedahmed commented:
@dan-f Yeah we can got to ```edx-domain/sudo```, I don't know what is expected behaviour here.

@cpennington , @explorerleslie ?

waheedahmed commented:
@dan-f Addressed all your comments, please take a look.

dan-f commented:
> @dan-f I can't reproduce the above issue on instructor dashboard data download section, network tab is not showing any redirection to sudo.
> May be your django-sudo session was expired.

@waheedahmed my django-sudo session did expire, and that's my point. Based on your implementation, If I'm looking at the instructor dashboard with an expired session I would expect that clicking on any button would result in my being redirected to the django-sudo login page. The redirects are not working, however. @explorerleslie this was my main question for you - what do you expect the behavior to be?

dan-f commented on common/test/acceptance/pages/lms/instructor_dashboard.py:
```diff
@@ -773,127 +774,182 @@ def is_browser_on_page(self):
 
     @property
-    def student_email_input(self):
+    def entrance_exam_student_email_input(self):
```
Are you still getting `invalid-name` pylint errors? It's probably because some of these names are getting long. You can append `# pylint: disable=invalid-name` to the line to prevent these errors. In general, I think we tend to avoid abbreviated names where possible in our platform.

dan-f commented:
@waheedahmed Thanks for responding to my comments. My only remaining question is regarding the expected behavior of the instructor dashboard, as described above. Once we resolve that, I'll give my thumbs. Thanks!

dan-f commented:
Also - I should mention: once we decide on how to handle that particular use case, I'd like to see a bok_choy test added.

explorerleslie commented:
@waheedahmed Dan and I just discussed this. My expectation was, when you've been auto-logged out of sudo mode, and you click a button that requires sudo, you're redirected to the sudo login page. Right now, nothing happens when you click, which is extremely confusing for the user who won't know why nothing is happening.

waheedahmed commented:
@explorerleslie, @dan-f OK great, I'll let you know after doing this.

waheedahmed commented:
@dan-f, @explorerleslie I implemented the required behaviour for data download tab of instructor dashboard and updated the sandbox, can you please check it looks good or not?

And one thing more I implemented this for ```list_instructor_task``` too on which a timer is set which actually requests every 20 seconds and get list of pending instructor tasks, so when a users session will expire it will automatically redirect it to sudo page.

explorerleslie commented:
@waheedahmed what is the timeout set to on the sandbox? I can't seem to ever get the Data Downloads page to time out to see what happens when I try to generate a report after the timeout.

dan-f commented on lms/static/coffee/src/instructor_dashboard/data_download.coffee:
```diff
@@ -140,6 +148,8 @@ class DataDownload
         url: url
         error: (std_ajax_err) =>
+          if std_ajax_err.status == 401
+            window.location.reload()
```
Please extract a helper method rather than copy and paste these lines.

dan-f replied:
Additionally - what is the reason for special-casing the JSON version of sudo-required with a `401` instead of the standard redirect?

waheedahmed replied:
@dan-f Because we can't handle redirect in javascript, browser always follow the redirected url but we need to check if response.status is 302 then reload the ```window.location```.

waheedahmed commented:
test this please

waheedahmed commented:
@explorerleslie I think its by default 3 hours: https://github.com/edx/django-sudo/blob/master/sudo/settings.py#L20

@cpennington FYI.

waheedahmed commented:
@explorerleslie I'm going on vacations for two weeks and now only this change is remaining, if you want this to be merged as soon as possible I talked with @awaisdar001 about this and he is willing to take this from here if you want or it can wait until I'm back? 

There is only javascript change is remaining for whole instructor dashboard.

waheedahmed commented:
test this please

explorerleslie commented:
@waheedahmed thanks. If @awaisdar001 can take this over, that would be great. @awaisdar001 I'm having a lot of trouble testing this because the timeout is 3 hours, can you please make the timeout some short time, like 2 mins, on this sandbox?

awaisdar001 commented:
@explorerleslie You can also manually delete specific cookies from the browser to test this.  That way you can expire the session whenever you want. What do you think?
I can find a way for you to do that.

explorerleslie commented:
@awaisdar001 I *think* it would be easiest to create a one-line change in this PR that you don't merge overriding the timeout time, but whatever will make it easiest for me to just test. :)

dan-f commented:
@awaisdar001 - I'd like to echo @explorerleslie's point regarding the difficulty to test the sandbox. I think it would be best for us all if you can change the timeout setting on the server to a very short value for testing.

Moreover, how long do we want the timeout to be in production? Until recently I did not realize that we were just using the built-in default timeout. Does that default value suit our needs?

awaisdar001 commented:
I couldn't find the `django.sudo.m.sandbox.edx.org` work for me, It might have expired or somthing, I have created a new sandbox with the timeout limit set to 2mints. 

http://dj-sudo.m.sandbox.edx.org
Feel free to test;
FYI @explorerleslie  @dan-f

awaisdar001 commented:
@dan-f The default timeout is `10800 seconds` (3 hours).

awaisdar001 commented:
Updated the Ajax requests to redirect the user to `sudo-page` when the session is expired and server returns `401`

dan-f commented:
Who is owning the value of the timeout - @explorerleslie @cpennington - I am not sure if this is a product or engineering decision, but the default value of 3 hours seems... large to me.

dan-f commented:
Thanks @awaisdar001 for the update. I will begin re-testing now.

dan-f commented:
@awaisdar001 - I'm personally happy with the new behavior, but I added some comments to your last commit which I'd like to see get addressed. Please let me know when you address them.

awaisdar001 commented:
@dan-f I have addressed your comments, moved page reload code to the helper function.

explorerleslie commented:
:+1: from product on the behavior. wait for @dan-f to thumb before merging.

dan-f commented:
@awaisdar001 - other than my final nit (which I posted on the last commit), :+1:. Thanks for picking this up and addressing my comments! Don't forget to squash your commits.

awaisdar001 commented:
@dan-f Comments addressed and commits are `squash`

awaisdar001 commented:
@adampalay So we are ready to merge this fella?

adampalay commented:
woohoo!
